### DISCUSSION TITLE
Do I need to memoize selectors?

### DISCUSSION THREAD

**Discussion by mfbx9da4** [pmndrs/zustand#658] [Q&A] [answered]
> It is generally recommended to memoize selectors with useCallback. 

Is it generally recommended to memoize all selectors? Seems like a premature optimisation for the most part? 

If it's not a premature optimization seems like a major omission to exclude factories from the zustand library. Will it be death by a thousand cuts?

If most of the time it is a premature optimisation, under what circumstances is it advised?

**Comment by dai-shi** [comment, marked as answer]
That section is introduced in v3, because at that time we were expecting `useMutableSource` which would require memoized selectors (without stable selectors, it caused infinite loop). #422 

Now, we will have `useSyncExternalStore` instead and memoized selectors are not necessary. That section is removed in v4. #550 (Hmm, we can remove it now, but v4 will probably be released any time soon.)

**Reply by mfbx9da4** [reply]
Thanks for the response. Sounds like there are no performance considerations wrt necessity of memoising selectors. Could you confirm this please?

**Reply by dai-shi** [reply]
In the current implementation, having a stable selector will skip invoking the selector, so it's technically efficient to define a selector outside component. But, yes, it's premature optimization and will not be noticeable in practice.
